File: 4. Median of Two Sorted Arrays/Solution.kt
/**
 * Runtime: 252 ms, faster than 95.13% of Kotlin online submissions for Median of Two Sorted Arrays.
 * Memory Usage: 45.9 MB, less than 61.69% of Kotlin online submissions for Median of Two Sorted Arrays.
 */
class Solution {
    fun findMedianSortedArrays(nums1: IntArray, nums2: IntArray): Double {
        var nums1length = nums1.size;
        var nums2length = nums2.size;
        var numsLength = nums1length + nums2length;
        var halfIndex = Math.round((numsLength / 2).toDouble()).toInt();
        
        var medianIndexes: Array<Int>;
        if (numsLength % 2 == 0) {
            medianIndexes = arrayOf(((numsLength / 2) - 1).toInt(), (numsLength / 2).toInt());
        } else {
            medianIndexes = arrayOf(halfIndex);
        }
        
        if (nums1length == 0) {
             return getMedian(nums2, medianIndexes);
        }
        
        if (nums2length == 0) {
             return getMedian(nums1, medianIndexes);
        }
        
        var nums2currentIndex = 0;
        var mergedArray = IntArray(numsLength);
        for (i in 0..nums1length-1) {
            for (j in nums2currentIndex..nums2length-1) {
                if (nums2[j] <= nums1[i]) {
                    mergedArray[i + nums2currentIndex] = nums2[j];
                    nums2currentIndex++;
                } else {
                    break;
                }
            }
            
            mergedArray[i + nums2currentIndex] = nums1[i];
            
            if (i + nums2currentIndex > halfIndex) {
                break;
            }
        }
        
        for (j in nums2currentIndex..nums2length-1) {
            mergedArray[nums1length + nums2currentIndex] = nums2[j];
            nums2currentIndex++;
            if (nums1length + nums2currentIndex > halfIndex) {
                break;
            }
        }
        
        return getMedian(mergedArray, medianIndexes);
    }
    
    fun getMedian(array: IntArray, medianIndexes: Array<Int>): Double {
        var sum = 0;
        
        medianIndexes.forEach {
            sum += array[it];
        }
        
        return sum.toDouble() / medianIndexes.size.toDouble();
    }
}

File: 4. Median of Two Sorted Arrays/Solution.php
<?php

/**
 * Runtime: 32 ms, faster than 86.96% of PHP online submissions for Median of Two Sorted Arrays.
 * Memory Usage: 15.8 MB, less than 95.65% of PHP online submissions for Median of Two Sorted Arrays.
 */
class Solution {

    /**
     * @param Integer[] $nums1
     * @param Integer[] $nums2
     * @return Float
     */
    function findMedianSortedArrays($nums1, $nums2)
    {
        $nums1length = count($nums1);
        $nums2length = count($nums2);
        
        $numsLength = $nums1length + $nums2length;
        
        $medianIndexes = ($nums1length + $nums2length) % 2 === 0
                          ? [(int)($numsLength / 2) - 1, (int)($numsLength / 2)]
                          : [(int)round($numsLength / 2, 0, PHP_ROUND_HALF_DOWN)];
        
        if ($nums1length === 0) {
            return $this->findMedian($nums2, $medianIndexes);
        }
        
        if ($nums2length === 0) {
            return $this->findMedian($nums1, $medianIndexes);
        }
        
        $nums2currentIndex = 0;
        $mergedArray = [];
        for ($i = 0; $i < count($nums1); $i++) {
            for ($j = $nums2currentIndex; $j < count($nums2); $j++) {
                if ($nums2[$j] <= $nums1[$i]) {
                    $mergedArray[] = $nums2[$j];
                    $nums2currentIndex++;
                } else {
                    break;
                }
            }
            
            $mergedArray[] = $nums1[$i];
            
            if (count($mergedArray) > round($numsLength / 2, 0, PHP_ROUND_HALF_UP) + 1) {
                break;
            }
        }
        
        for ($j = $nums2currentIndex; $j < count($nums2); $j++) {
            $mergedArray[] = $nums2[$j];
            if (count($mergedArray) > round($numsLength / 2, 0, PHP_ROUND_HALF_UP) + 1) {
                break;
            }
        }
        
        return $this->findMedian($mergedArray, $medianIndexes);
    }
    
    function findMedian($array, $medianIndexes)
    {
        $sum = 0;
        foreach ($medianIndexes as $index) {
            $sum += $array[$index];
        }
        
        return $sum / count($medianIndexes);
    }
}